h/security/permission_map.py

```diff
@@ -0,0 +1,155 @@
+    Permission.Annotation.READ_REALTIME_UPDATES: [
+        [annotation_not_shared, annotation_created_by_user],
+        # For shared annotations the permissions are copied from the group
+        # of the annotation. We bend the predicate system here a little and
+        # put a raw permission in which will cause us to look up that
+        # permission.
+        [annotation_shared, Permission.Group.READ],
+    ],
+    Permission.Annotation.READ: [
+        [annotation_live, annotation_not_shared, annotation_created_by_user],
+        [annotation_live, annotation_shared, Permission.Group.READ],
+    ],
+    Permission.Annotation.FLAG: [
+        [annotation_live, annotation_not_shared, annotation_created_by_user],
+        [annotation_live, annotation_shared, Permission.Group.FLAG],
+    ],
+    Permission.Annotation.MODERATE: [
+        [annotation_live, annotation_shared, Permission.Group.MODERATE]
+    ],
+    # The user who created the annotation always has the these permissions
+    Permission.Annotation.UPDATE: [[annotation_live, annotation_created_by_user]],
+    Permission.Annotation.DELETE: [[annotation_live, annotation_created_by_user]],
+}
+
+# Predicates can define parents which they need to be true before they are
+# true. We don't put all the parents here because it would be fiddly and make
+# it harder to read.
+
+# This turns the abstract predicates above into lists which include all of
+# their parents in the correct order to evaluate them.
+PERMISSION_MAP = resolve_predicates(PERMISSION_MAP)
+
+
+def permits(identity: Identity, context, permission) -> bool:
+    """
+    Check whether a given identity has permission to operate on a context.
+
+    For example the identity might include a user, and the context a group
+    and the permission might ask whether the user can edit that group.
+
+    :param identity: Identity object of the user
+    :param context: Context object representing the objects acted upon
+    :param permission: Permission requested
+    """
+    if clauses := PERMISSION_MAP.get(permission):
+        for clause in clauses:
+            if _clause_true(clause, identity, context):
+                return True
+
+        return False
+
+    return False
+
+
+def _clause_true(clause, identity, context):
+    """Check whether every item in a clause is true."""
+
+    # There are lots of optimisations you could do with this (like caching
+    # predicate results), but at the time of writing our predicates are so fast
+    # to evaluate it generally makes it slower to try.
+    for predicate in clause:
+        try:
+            if not predicate(identity, context):
+                return False
+
+        # If the "predicate" isn't callable, we'll assume it's a permission
+        # and work out if we have that permission
+        except TypeError:
+            if not permits(identity, context, predicate):
+                return False
+
+    return True
```

Comment:
A `_predicate_true()` helper might actually be nicer than `_clause_true()` because then the body code of `permits()` could more clearly expose the overall logic. A permission is granted if, for **any** of the permission's clauses, **all** of the predicates in that clause are truthy:

```python
def permits(identity, context, permission):
    clauses = PERMISSION_MAP.get(permission)

    return any(
        all([_predicate_true(predicate, identity, context) for predicate in clause])
        for clause in clauses
    )
```

Or, maybe both helpers is clearer:

```python
def permits(identity, context, permission):
    return any([_clause_true(clause, identity, context)] for clause in PERMISSION_MAP.get(permission)])

def _clause_true(clause, identity, context):
    return all([_predicate_true(predicate, identity, context) for predicate in clause])

def _predicate_true(predicate, identity, context):
    ...
```

Reply:
I've used a `_predicate_true` and used any and all.